[BUG] Parser/Linter error: `beresp.ttl` cannot be assigned `parse_time_delta` response

**Describe the problem**

`set beresp.ttl = parse_time_delta(beresp.http.Edge-Control:cache-maxage);` causes lint error due to type mismatch as follows.

```sh
🔥 [ERROR] beresp.ttl wants type RTIME but assign INTEGER
```

This snippet is quoted from the example of the `parse_time_delta` function document.
https://developer.fastly.com/reference/vcl/functions/date-and-time/parse-time-delta/#example

`beresp.ttl` is actually `RTIME` type in the [document](https://developer.fastly.com/reference/vcl/variables/backend-response/beresp-ttl/).
`parse_time_delta` is `INTEGER` type ([document](https://developer.fastly.com/reference/vcl/functions/date-and-time/parse-time-delta/)), so falco is valid implementation for the document.

Therefore, type mismatch seems intentional behavior, however, the code snippet above is actually the valid and works properly.

idk why the code is valid and whether it truly follows the vcl spec or not. but it seems like a bug.

**VCL code that cause the problem**

```vcl
sub vcl_fetch {
    if (parse_time_delta(beresp.http.Edge-Control:cache-maxage) >= 0) {
      set beresp.ttl = parse_time_delta(beresp.http.Edge-Control:cache-maxage);
    }
    return(deliver);
}
```

**Screenshots**

![image](https://user-images.githubusercontent.com/14837380/148173111-808acd3b-3996-4c88-b601-57509f0d43c2.png)

**Expected behavior**

The test should pass.

**Desktop (please complete the following information):**
 - OS: Ubuntu (WSL2)
 - Shell: bash

--- a/linter/errors.go
+++ b/linter/errors.go
@@ -162,21 +162,17 @@ func ErrorCodeRange(m *ast.Meta, code int64) *LintError {
 	}
 }
 
-// FIXME: accept *ast.Meta
-func InvalidTypeOperator(m *ast.Meta, op string, expects ...types.Type) *LintError {
-	es := make([]string, len(expects))
-	for i, v := range expects {
-		es[i] = v.String()
-	}
-
+func InvalidTypeOperator(m *ast.Meta, op string, left, right types.Type) *LintError {
 	return &LintError{
 		Severity: ERROR,
 		Token:    m.Token,
-		Message:  fmt.Sprintf(`could not operand, "%s" operator expects type %s on right expression`, op, strings.Join(es, " or ")),
+		Message: fmt.Sprintf(
+			`invalid operator, "%s" count not use between left %s and %s right types`,
+			op, left, right,
+		),
 	}
 }
 
-// FIXME: accept *ast.Meta
 func InvalidOperator(m *ast.Meta, op string, left types.Type) *LintError {
 	return &LintError{
 		Severity: ERROR,

--- a/linter/linter.go
+++ b/linter/linter.go
@@ -651,225 +651,22 @@ func (l *Linter) lintSetStatement(stmt *ast.SetStatement, ctx *context.Context) 
 	// https://developer.fastly.com/reference/vcl/operators/#assignment-operators
 	//
 	// Above document is not enough to explain for other types... actually more complex type coparison may occur.
+	// We investigated type comparison and summarized.
+	// See: https://docs.google.com/spreadsheets/d/16xRPugw9ubKA1nXHIc5ysVZKokLLhysI-jAu3qbOFJ8/edit#gid=0
 	switch stmt.Operator.Operator {
-	case "+=":
-		l.lintAdditionOperator(stmt.Operator, left, right)
-	case "-=":
-		l.lintSubtractionOperator(stmt.Operator, left, right)
-	case "*=":
-		l.lintMutiplicationOperator(stmt.Operator, left, right)
-	case "/=":
-		l.lintDivistionOperator(stmt.Operator, left, right)
-	case "%=":
-		l.lintRemainderOperator(stmt.Operator, left, right)
-	case "|=":
-		l.lintBitwiseOROperator(stmt.Operator, left, right)
-	case "&=":
-		l.lintBitwiseANDOperator(stmt.Operator, left, right)
-	case "^=":
-		l.lintBitwiseXOROperator(stmt.Operator, left, right)
-	case "<<=":
-		l.lintLeftShiftOperator(stmt.Operator, left, right)
-	case ">>=":
-		l.lintRightShiftOperator(stmt.Operator, left, right)
-	case "rol=":
-		l.lintLeftRorateOperator(stmt.Operator, left, right)
-	case "ror=":
-		l.lintRightRotateOperator(stmt.Operator, left, right)
-	case "&&=":
-		l.lintLogicalANDOperator(stmt.Operator, left, right)
-	case "||=":
-		l.lintLogicalOROperator(stmt.Operator, left, right)
+	case "+=", "-=":
+		l.lintAddSubOperator(stmt.Operator, left, right, isLiteralExpression(stmt.Value))
+	case "*=", "/=", "%=":
+		l.lintArithmeticOpereator(stmt.Operator, left, right, isLiteralExpression(stmt.Value))
+	case "|=", "&=", "^=", "<<=", ">>=", "rol=", "ror=":
+		l.lintBitwiseOperator(stmt.Operator, left, right)
+	case "||=", "&&=":
+		l.lintLogicalOperator(stmt.Operator, left, right)
 	default: // "="
-		l.lintAssignOperator(stmt.Value, stmt.Ident.Value, left, right)
+		l.lintAssignOperator(stmt.Operator, stmt.Ident.Value, left, right, isLiteralExpression(stmt.Value))
 	}
 
 	return types.NeverType
-}
-
-func (l *Linter) lintAssignOperator(value ast.Expression, name string, left, right types.Type) {
-	switch left {
-	case types.StringType:
-		switch right {
-		case types.StringType:
-			return
-		case types.FloatType, types.RTimeType, types.IntegerType, types.BoolType, types.NullType:
-			// Note: on right expression of set statement, could not assign with literal,
-			// but can assign with variable/identity as stringified value
-			if isLiteralExpression(value) {
-				l.Error(InvalidType(value.GetMeta(), name, left, right))
-			}
-		default:
-			l.Error(InvalidType(value.GetMeta(), name, left, right))
-		}
-	default:
-		if left != right {
-			l.Error(InvalidType(value.GetMeta(), name, left, right))
-		}
-	}
-}
-
-func (l *Linter) lintAdditionOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType, types.FloatType, types.RTimeType:
-		if !expectType(right, types.FloatType, types.RTimeType, types.IntegerType) {
-			l.Error(InvalidTypeOperator(
-				op.Meta, op.Operator, types.IntegerType, types.FloatType, types.RTimeType,
-			).Match(OPERATOR_ASSIGNMENT))
-		}
-	case types.BoolType, types.AclType, types.BackendType:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-	// "+=" operator can use for STRING type to concat.
-}
-
-func (l *Linter) lintSubtractionOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType, types.FloatType, types.RTimeType:
-		if !expectType(right, types.FloatType, types.RTimeType, types.IntegerType) {
-			l.Error(InvalidTypeOperator(
-				op.Meta, op.Operator, types.IntegerType, types.FloatType, types.RTimeType,
-			).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintMutiplicationOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType, types.FloatType, types.RTimeType:
-		if !expectType(right, types.FloatType, types.IntegerType) {
-			l.Error(InvalidTypeOperator(
-				op.Meta, op.Operator, types.IntegerType, types.FloatType,
-			).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintDivistionOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType, types.FloatType, types.RTimeType:
-		if !expectType(right, types.FloatType, types.IntegerType) {
-			l.Error(InvalidTypeOperator(
-				op.Meta, op.Operator, types.IntegerType, types.FloatType,
-			).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintRemainderOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType, types.FloatType, types.RTimeType:
-		if !expectType(right, types.FloatType, types.IntegerType) {
-			l.Error(InvalidTypeOperator(
-				op.Meta, op.Operator, types.IntegerType, types.FloatType,
-			).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintBitwiseOROperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType:
-		if !expectType(right, types.IntegerType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.IntegerType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintBitwiseANDOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType:
-		if !expectType(right, types.IntegerType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.IntegerType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-func (l *Linter) lintBitwiseXOROperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType:
-		if !expectType(right, types.IntegerType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.IntegerType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintLeftShiftOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType:
-		if !expectType(right, types.IntegerType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.IntegerType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-func (l *Linter) lintRightShiftOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType:
-		if !expectType(right, types.IntegerType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.IntegerType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintLeftRorateOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType:
-		if !expectType(right, types.IntegerType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.IntegerType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintRightRotateOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.IntegerType:
-		if !expectType(right, types.IntegerType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.IntegerType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintLogicalANDOperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.BoolType:
-		if !expectType(right, types.BoolType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.BoolType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
-}
-
-func (l *Linter) lintLogicalOROperator(op *ast.Operator, left, right types.Type) {
-	switch left {
-	case types.BoolType:
-		if !expectType(right, types.BoolType) {
-			l.Error(InvalidTypeOperator(op.Meta, op.Operator, types.BoolType).Match(OPERATOR_ASSIGNMENT))
-		}
-	default:
-		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
-	}
 }
 
 func (l *Linter) lintUnsetStatement(stmt *ast.UnsetStatement, ctx *context.Context) types.Type {
@@ -1035,7 +832,7 @@ func (l *Linter) lintAddStatement(stmt *ast.AddStatement, ctx *context.Context) 
 		}
 		l.Error(err.Match(OPERATOR_ASSIGNMENT))
 	}
-	l.lintAssignOperator(stmt.Value, stmt.Ident.Value, left, right)
+	l.lintAssignOperator(stmt.Operator, stmt.Ident.Value, left, right, isLiteralExpression(stmt.Value))
 
 	return types.NeverType
 }

--- a/linter/linter_test.go
+++ b/linter/linter_test.go
@@ -1465,3 +1465,19 @@ sub vcl_recv {
 		}
 	})
 }
+
+// https://github.com/ysugimoto/falco/issues/39
+func TestPassIssue39(t *testing.T) {
+	t.Run("pass", func(t *testing.T) {
+		input := `
+sub vcl_fetch {
+	### FASTLY fetch
+    if (parse_time_delta(beresp.http.Edge-Control:cache-maxage) >= 0) {
+      set beresp.ttl = parse_time_delta(beresp.http.Edge-Control:cache-maxage);
+    }
+    return(deliver);
+}
+`
+		assertNoError(t, input)
+	})
+}

--- a/linter/operator.go
+++ b/linter/operator.go
@@ -1,0 +1,202 @@
+package linter
+
+import (
+	"github.com/ysugimoto/falco/ast"
+	"github.com/ysugimoto/falco/types"
+)
+
+// See expression type comparison table:
+// https://docs.google.com/spreadsheets/d/16xRPugw9ubKA1nXHIc5ysVZKokLLhysI-jAu3qbOFJ8/edit#gid=0
+
+// Lint assignment operator of "="
+func (l *Linter) lintAssignOperator(op *ast.Operator, name string, left, right types.Type, isLiteral bool) {
+	switch left {
+	case types.IntegerType:
+		switch right {
+		// allows both variable and literal
+		case types.IntegerType:
+			return
+		// allows variable only, disallow literal
+		case types.FloatType, types.RTimeType, types.TimeType:
+			if isLiteral {
+				l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+			}
+		// disallow
+		default:
+			l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+		}
+	case types.FloatType:
+		switch right {
+		// allows both variable and literal
+		case types.IntegerType, types.FloatType:
+			return
+		// allows variable only, disallow literal
+		case types.RTimeType, types.TimeType:
+			if isLiteral {
+				l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+			}
+		// disallow
+		default:
+			l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+		}
+	case types.StringType:
+		switch right {
+		// allows both variable and literal
+		case types.StringType, types.BoolType:
+			return
+		// allows variable only, disallow literal
+		case types.IntegerType, types.FloatType, types.RTimeType, types.TimeType, types.IPType:
+			if isLiteral {
+				l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+			}
+		// disallow
+		default:
+			l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+		}
+	case types.RTimeType, types.TimeType:
+		switch right {
+		// allows both variable and literal
+		case types.RTimeType, types.TimeType:
+			return
+		// allows variable only, disallow literal
+		case types.IntegerType, types.FloatType:
+			if isLiteral {
+				l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+			}
+		// disallow
+		default:
+			l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+		}
+	case types.IPType:
+		switch right {
+		// allows both variable and literal
+		case types.StringType, types.IPType:
+			return
+		// disallow
+		default:
+			l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+		}
+	default: // types.BackendType or types.BoolType
+		if left != right {
+			l.Error(InvalidType(op.Meta, name, left, right).Match(OPERATOR_ASSIGNMENT))
+		}
+	}
+}
+
+// Lint addition and subtraction operator of "+=" and "-="
+func (l *Linter) lintAddSubOperator(op *ast.Operator, left, right types.Type, isLiteral bool) {
+	switch left {
+	case types.IntegerType:
+		switch right {
+		// allows both variable and literal
+		case types.IntegerType:
+			return
+		// allows variable only, disallow literal
+		case types.FloatType, types.RTimeType, types.TimeType:
+			if isLiteral {
+				l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+			}
+		// disallow
+		default:
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	case types.FloatType:
+		switch right {
+		// allows both variable and literal
+		case types.IntegerType, types.FloatType:
+			return
+		// allows variable only, disallow literal
+		case types.RTimeType, types.TimeType:
+			if isLiteral {
+				l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+			}
+		// disallow
+		default:
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	case types.RTimeType:
+		switch right {
+		// allows both variable and literal
+		case types.RTimeType:
+			return
+		// allows variable only, disallow literal
+		case types.IntegerType, types.FloatType, types.TimeType:
+			if isLiteral {
+				l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+			}
+		// disallow
+		default:
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	case types.TimeType:
+		switch right {
+		// allows both variable and literal
+		case types.RTimeType:
+			return
+		// allows variable only, disallow literal
+		case types.IntegerType, types.FloatType:
+			if isLiteral {
+				l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+			}
+		// disallow
+		default:
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	default:
+		// disallow other types
+		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_CONDITIONAL))
+	}
+}
+
+// Lint arithmetic operator excepts addition and subtraction of "*=", "/=" and "%="
+func (l *Linter) lintArithmeticOpereator(op *ast.Operator, left, right types.Type, isLiteral bool) {
+	switch left {
+	case types.IntegerType:
+		switch right {
+		// allows both variable and literal
+		case types.IntegerType:
+			return
+		// allows variable only, disallow literal
+		case types.FloatType:
+			if isLiteral {
+				l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+			}
+		// disallow
+		default:
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	case types.FloatType, types.RTimeType:
+		switch right {
+		// allows both variable and literal
+		case types.IntegerType, types.FloatType:
+			return
+		// disallow
+		default:
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	default:
+		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_ASSIGNMENT))
+	}
+}
+
+func (l *Linter) lintBitwiseOperator(op *ast.Operator, left, right types.Type) {
+	switch left {
+	case types.IntegerType:
+		if right != types.IntegerType {
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	default:
+		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_CONDITIONAL))
+	}
+}
+
+func (l *Linter) lintLogicalOperator(op *ast.Operator, left, right types.Type) {
+	switch left {
+	case types.BoolType:
+		if right != types.BoolType {
+			l.Error(InvalidTypeOperator(op.Meta, op.Operator, left, right).Match(OPERATOR_CONDITIONAL))
+		}
+	default:
+		l.Error(InvalidOperator(op.Meta, op.Operator, left).Match(OPERATOR_CONDITIONAL))
+	}
+}

--- a/linter/operator.go
+++ b/linter/operator.go
@@ -148,7 +148,7 @@ func (l *Linter) lintAddSubOperator(op *ast.Operator, left, right types.Type, is
 	}
 }
 
-// Lint arithmetic operator excepts addition and subtraction of "*=", "/=" and "%="
+// Lint arithmetic operator excepts addition and subtraction. Lint "*=", "/=" and "%=" operator
 func (l *Linter) lintArithmeticOpereator(op *ast.Operator, left, right types.Type, isLiteral bool) {
 	switch left {
 	case types.IntegerType:
@@ -179,6 +179,7 @@ func (l *Linter) lintArithmeticOpereator(op *ast.Operator, left, right types.Typ
 	}
 }
 
+// Lint bitwise related operators "|=", "&=", "^=", "<<=", ">>=", "rol=" and "ror=".
 func (l *Linter) lintBitwiseOperator(op *ast.Operator, left, right types.Type) {
 	switch left {
 	case types.IntegerType:
@@ -190,6 +191,7 @@ func (l *Linter) lintBitwiseOperator(op *ast.Operator, left, right types.Type) {
 	}
 }
 
+// Lint logical operators "||=" and "&&=".
 func (l *Linter) lintLogicalOperator(op *ast.Operator, left, right types.Type) {
 	switch left {
 	case types.BoolType:

